consensus/src/consensus/inner.rs

```diff
@@ -0,0 +1,508 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::sync::Arc;
+
+use crate::{
+    error::ConsensusError,
+    Consensus,
+    ConsensusParameters,
+    CreatePartialTransactionRequest,
+    DeserializedLedger,
+    DynLedger,
+    MemoryPool,
+    Tx,
+};
+use anyhow::*;
+use snarkos_storage::{
+    BlockStatus,
+    Digest,
+    ForkDescription,
+    SerialBlock,
+    SerialTransaction,
+    Storage,
+    VMRecord,
+    VMTransaction,
+};
+use snarkvm_dpc::{
+    testnet1::{
+        instantiated::{Components, InstantiatedDPC},
+        payload::Payload as RecordPayload,
+        transaction::amount::AleoAmount,
+        Record as DPCRecord,
+        TransactionKernel,
+    },
+    DPCScheme,
+};
+use snarkvm_posw::txids_to_roots;
+use snarkvm_utilities::{to_bytes, ToBytes};
+use std::io::Result as IoResult;
+use tokio::sync::mpsc;
+
+use snarkos_metrics::{self as metrics, misc::*};
+
+use rand::thread_rng;
+
+use super::message::{ConsensusMessage, CreateTransactionRequest, TransactionResponse};
+
+pub struct ConsensusInner {
+    pub public: Arc<Consensus>,
+    pub ledger: DynLedger,
+    pub memory_pool: MemoryPool,
+    pub storage: Arc<dyn Storage>,
+}
+
+impl ConsensusInner {
+    fn receive_transaction(&mut self, transaction: Box<SerialTransaction>) -> bool {
+        if transaction.value_balance.is_negative() {
+            error!("Received a transaction that was a coinbase transaction");
+            return false;
+        }
+        match self.verify_transaction(&*transaction) {
+            Ok(true) => (),
+            Ok(false) => {
+                error!("Received a transaction that was invalid");
+                return false;
+            }
+            Err(e) => {
+                error!(
+                    "failed to validate transaction -- note this does not mean the transaction was valid or invalid: {:?}",
```

Comment:
nit: the method's name is `verify_transaction`, so "failed to verify" would be more fitting here and in other such places